sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/nodes/DebugExprVarNode.java

```diff
@@ -0,0 +1,95 @@
+ * OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
+ * GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.oracle.truffle.llvm.runtime.debug.debugexpr.nodes;
+
+import com.oracle.truffle.api.Scope;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.interop.InvalidArrayIndexException;
+import com.oracle.truffle.api.interop.UnknownIdentifierException;
+import com.oracle.truffle.api.interop.UnsupportedMessageException;
+import com.oracle.truffle.llvm.runtime.LLVMLanguage;
+import com.oracle.truffle.llvm.runtime.debug.LLVMDebuggerValue;
+import com.oracle.truffle.llvm.runtime.debug.debugexpr.parser.Parser;
+import com.oracle.truffle.llvm.runtime.nodes.api.LLVMExpressionNode;
+
+public class DebugExprVarNode extends LLVMExpressionNode {
+
+    private final String name;
+    private Iterable<Scope> scopes;
+
+    public DebugExprVarNode(String name, Iterable<Scope> scopes) {
+        this.name = name;
+        this.scopes = scopes;
+    }
+
+    @Override
+    public Object executeGeneric(VirtualFrame frame) {
+        LLVMLanguage.getLLVMContextReference().get();
+        for (Scope scope : scopes) {
+            Object vars = scope.getVariables();
+            InteropLibrary library = InteropLibrary.getFactory().getUncached();
+            library.hasMembers(vars);
+            try {
+                final Object memberKeys = library.getMembers(vars);
+                library.hasArrayElements(memberKeys);
+                for (long i = 0; i < library.getArraySize(memberKeys); i++) {
+                    final String memberKey = (String) library.readArrayElement(memberKeys, i);
+                    if (!memberKey.equals(name))
+                        continue;
+                    Object member = library.readMember(vars, memberKey);
+                    LLVMDebuggerValue ldv = (LLVMDebuggerValue) member;
+                    Object metaObj = ldv.getMetaObject();
+                    if (metaObj == null) {
+                        return member;
+                    }
+                    // System.out.println(name + "|metaObj.toString() = " + metaObj.toString());
+                    // System.out.println(name + "|ldv.toString() = " + ldv.toString());
+                    if (metaObj.toString().contentEquals("int")) {
```

Comment:
Languages other than C may not use these strings as type names. The `metaObj` for primitive values from any language is a `LLVMSourceBasicType`, which exposes `getKind` and `getSize` methods that you can use to determine how to parse the value. Please know that `getSize` in this case is the bit-size of the type, and that in C you can easily have integer types that are not the usual power-of-two sizes. In that case you can use `if (type.getSize() < Long.SIZE) { Long.parseLong(str) } else if (type.getSize() < Integer.SIZE) { ...` to choose the correct representation (Widening the types in this case is fine with me).